TPC/TPCcalib/AliTPCcalibAlign.cxx

```diff
@@ -715,10 +715,13 @@ void  AliTPCcalibAlign::ExportTrackPoints(AliVEvent *event){
     AliTPCseed *seed0 = 0,*seed1=0;
     AliTPCseed tpcSeed0, tpcSeed1;
     //
+    Int_t nFriendTracks = Vfriend->GetNumberOfTracks();
+    if (index0 > nFriendTracks) continue; // most likely, we didn't save the friends for this event
     friendTrack = const_cast<AliVfriendTrack*>(Vfriend->GetTrack(index0));
     if (!friendTrack) continue;
     if (friendTrack->GetTPCseed(tpcSeed0)==0) seed0=&tpcSeed0;
     if (index1>0){
+      if (index1 > nFriendTracks) continue; // most likely, we didn't save the friends for this event
```

Comment:
same comment as above

Reply:
Hi @shahor02, I don't see how that would be dangerous? I thought the friend tree has to have the same number of entries to preserve the matching - so some of them will be empty and you get a null pointer for those cases, the same as when calling track->GetFriendTrack() (which just falls back to reading the same entry from the friend tree/branch).

Reply:
Hello,

Meanwhile Haavard approved the PR, so we will need a new commit in case we want to change this.

Chiara

Reply:
I think we will want to change this, best to have a TPC expert look at this, e.g. @miranov25

Reply:
Ciao Mikolaj.
For me it is fine to change it, for now this works, even if it is probably not the most elegant. 
Chiara

Reply:
Hi, today I've fixed many lines like that, but it seems missed this one. What I've changed is from
````
AliVTrack *track = event->GetVTrack(i);
...
friendTrack = const_cast<AliVfriendTrack*>(Vfriend->GetTrack(i));
````
to 
````
AliVTrack *track = event->GetVTrack(i1);
AliVfriendTrack *friendTrack = 0;
if (track->IsA()==AliESDtrack::Class()) {
   friendTrack = (AliVfriendTrack*)(((AliESDtrack*)track)->GetFriendTrack());
}
else {
   friendTrack = const_cast<AliVfriendTrack*>(vFriend->GetTrack(i));
}
````

Shall I change this file too?

Reply:
Hi Ruben, sure, go ahead, but we were hoping to avoid these kinds of checks (again introducing different code paths based on input type). In principle we could think about introducing something like this in the V interface and flat ESD, but that might be too much work - including some CPU overhead due to the need to explicitly attach some pointers/indexes to the tracks (happens in the AliAnalysisManager for offline, always introducing this overhead). I think index based access to the friend tree directly is a much more elegant solution.
Now about the check: it is wrong algorithmically, it will possibly bias the result as it will only process the friends attached to the first 3000(fMax...) tracks only. Bias might be small, but still... :)

Reply:
@mkrzewic : I agree about the indixed access, but for historical reasons we have to support the esdTrack->GetFriendTrack method. One can access sparse friends also via indices:
````
int idx = esdTrack->GetFriendTrackID();
AliESDfriendTrack* frTrack = idx<0 ? 0 : esdFrEvent->GetTrack(idx);
````

Reply:
@mkrzewic : "Now about the check: it is wrong algorithmically, it will possibly bias the result as it will only process the friends attached to the first 3000(fMax...) tracks only. Bias might be small, but still... :)" --> 
Actually also the current code:

friendTrack = const_cast<AliVfriendTrack*>(Vfriend->GetTrack(index0));

is wrong, as it is not guaranteed that the index of a track is not the same as the index of its friend.

Reply:
done as https://github.com/alisw/AliRoot/pull/593

Reply:
the joy of having friends...

Reply:
wait until we include enemies in AliRoot... :)

Reply:
LOL